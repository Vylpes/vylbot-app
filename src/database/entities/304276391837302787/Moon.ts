import { Column, Entity, IsNull } from "typeorm";
import BaseEntity from "../../../contracts/BaseEntity";
import AppDataSource from "../../dataSources/appDataSource";

@Entity()
export default class Moon extends BaseEntity {
    constructor(moonNumber: number, description: string, userId: string) {
        super();

        this.MoonNumber = moonNumber;
        this.Description = description;
        this.UserId = userId;
    }

    @Column()
    MoonNumber: number;

    @Column()
    Description: string;

    @Column({ nullable: true })
    WhenArchived?: Date;

    @Column()
    UserId: string;

    public static async FetchMoonsByUserId(userId: string): Promise<Moon[] | null> {
        const repository = AppDataSource.getRepository(Moon);

        const all = await repository.find({ where: { UserId: userId } });

        return all;
    }

    public static async FetchPaginatedMoonsByUserId(userId: string, pageLength: number, page: number): Promise<[ Moon[], number ]> {
        const rangeStart = page * pageLength;

        const repository = AppDataSource.getRepository(Moon);

        const moons = await repository.findAndCount({
            where: { UserId: userId, WhenArchived: IsNull() },
            order: { MoonNumber: "ASC" },
            skip: rangeStart,
            take: pageLength,
        });

        return moons;
    }

    public static async FetchMoonCountByUserId(userId: string): Promise<number> {
        const repository = AppDataSource.getRepository(Moon);

        const count = await repository.count({ where: { UserId: userId } });

        return count;
    }
}
