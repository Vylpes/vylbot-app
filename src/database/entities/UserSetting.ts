import { Column, Entity} from "typeorm";
import AppDataSource from "../dataSources/appDataSource";
import BaseEntity from "../../contracts/BaseEntity";

@Entity()
export default class UserSetting extends BaseEntity {
    constructor(userId: string, key: string, value: string) {
        super();

        this.UserId = userId;
        this.Key = key;
        this.Value = value;
    }

    @Column()
    UserId: string;

    @Column()
    Key: string;

    @Column()
    Value: string;

    public UpdateValue(value: string) {
        this.Value = value;
    }

    public static async FetchOneByKey(userId: string, key: string, relations?: string[]): Promise<UserSetting | null> {
        const repository = AppDataSource.getRepository(UserSetting);

        const single = await repository.findOne({ where: { UserId: userId, Key: key }, relations: relations || {} });

        return single;
    }
}
