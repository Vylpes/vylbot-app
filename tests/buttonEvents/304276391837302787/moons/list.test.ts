import {ActionRowBuilder, ButtonBuilder, ButtonInteraction, EmbedBuilder} from "discord.js";
import List from "../../../../src/buttonEvents/304276391837302787/moons/list";
import UserSetting from "../../../../src/database/entities/UserSetting";
import Moon from "../../../../src/database/entities/304276391837302787/Moon";

describe("GIVEN happy flow", () => {
    let updatedWithEmbeds: EmbedBuilder[] | undefined;
    let updatedWithRows: ActionRowBuilder<ButtonBuilder>[] | undefined;

    const interaction = {
        guild: {
            members: {
                cache: {
                    find: jest.fn().mockReturnValue({
                        user: {
                            username: "username",
                        },
                    }),
                },
            },
        },
        reply: jest.fn(),
        update: jest.fn((options: any) => {
            updatedWithEmbeds = options.embeds;
            updatedWithRows = options.components;
        }),
        customId: "moons list userId 0",
    } as unknown as ButtonInteraction;

    beforeAll(async () => {
        UserSetting.FetchOneByKey = jest.fn();
        Moon.FetchPaginatedMoonsByUserId = jest.fn().mockResolvedValue([
            [
                {
                    MoonNumber: 1,
                    Description: "Test Description",
                }
            ],
            1,
        ]);

        await List(interaction);
    });

    test("EXPECT moons to be fetched", () => {
        expect(Moon.FetchPaginatedMoonsByUserId).toHaveBeenCalledTimes(1);
        expect(Moon.FetchPaginatedMoonsByUserId).toHaveBeenCalledWith("userId", 10, 0);
    });

    test("EXPECT interaction.update to be called", () => {
        expect(interaction.update).toHaveBeenCalledTimes(1);
    });

    test("EXPECT embed to be updated", () => {
        expect(updatedWithEmbeds).toMatchSnapshot();
    });

    test("EXPECT row to be updated", () => {
        expect(updatedWithRows).toMatchSnapshot();
    });
});

describe("GIVEN interaction.guild is null", () => {
    const interaction = {
        guild: null,
        reply: jest.fn(),
        update: jest.fn(),
    } as unknown as ButtonInteraction;

    beforeAll(async () => {
        UserSetting.FetchOneByKey = jest.fn();

        await List(interaction);
    });

    test("EXPECT function to return", () => {
        expect(interaction.reply).not.toHaveBeenCalled();
        expect(interaction.update).not.toHaveBeenCalled();
        expect(UserSetting.FetchOneByKey).not.toHaveBeenCalled();
    });
});

describe("GIVEN userId parameter is undefined", () => {
    const interaction = {
        guild: {},
        reply: jest.fn(),
        update: jest.fn(),
        customId: "moons list",
    } as unknown as ButtonInteraction;

    beforeAll(async () => {
        UserSetting.FetchOneByKey = jest.fn();

        await List(interaction);
    });

    test("EXPECT function to return", () => {
        expect(interaction.reply).not.toHaveBeenCalled();
        expect(interaction.update).not.toHaveBeenCalled();
        expect(UserSetting.FetchOneByKey).not.toHaveBeenCalled();
    });
});

describe("GIVEN page parameter is undefined", () => {
    const interaction = {
        guild: {},
        reply: jest.fn(),
        update: jest.fn(),
        customId: "moons list userId",
    } as unknown as ButtonInteraction;

    beforeAll(async () => {
        UserSetting.FetchOneByKey = jest.fn();

        await List(interaction);
    });

    test("EXPECT function to return", () => {
        expect(interaction.reply).not.toHaveBeenCalled();
        expect(interaction.update).not.toHaveBeenCalled();
        expect(UserSetting.FetchOneByKey).not.toHaveBeenCalled();
    });
});

describe("GIVEN no moons for the user is returned", () => {
    const interaction = {
        guild: {
            members: {
                cache: {
                    find: jest.fn().mockReturnValue({
                        user: {
                            username: "username",
                        },
                    }),
                },
            },
        },
        reply: jest.fn(),
        update: jest.fn(),
        customId: "moons list userId 0",
    } as unknown as ButtonInteraction;

    beforeAll(async () => {
        UserSetting.FetchOneByKey = jest.fn();
        Moon.FetchPaginatedMoonsByUserId = jest.fn().mockResolvedValue(undefined)

        await List(interaction);
    });

    test("EXPECT moons function to be called", () => {
        expect(Moon.FetchPaginatedMoonsByUserId).toHaveBeenCalledTimes(1);
    });

    test("EXPECT error replied", () => {
        expect(interaction.reply).toHaveBeenCalledTimes(1);
        expect(interaction.reply).toHaveBeenCalledWith("username does not have any moons or page is invalid.");
    });

    describe("GIVEN member is not in cache", () => {
        const interaction = {
            guild: {
                members: {
                    cache: {
                        find: jest.fn().mockReturnValue(undefined),
                    },
                    fetch: jest.fn().mockResolvedValue({
                        user: {
                            username: "username",
                        },
                    }),
                },
            },
            reply: jest.fn(),
            update: jest.fn(),
            customId: "moons list userId 0",
        } as unknown as ButtonInteraction;

        beforeAll(async () => {
            UserSetting.FetchOneByKey = jest.fn();
            Moon.FetchPaginatedMoonsByUserId = jest.fn().mockResolvedValue(undefined)

            await List(interaction);
        });

        test("EXPECT API to be called", () => {
            expect(interaction.guild?.members.fetch).toHaveBeenCalledTimes(1);
            expect(interaction.guild?.members.fetch).toHaveBeenCalledWith("userId");
        });

        test("EXPECT error replied with username", () => {
            expect(interaction.reply).toHaveBeenCalledTimes(1);
            expect(interaction.reply).toHaveBeenCalledWith("username does not have any moons or page is invalid.");
        });
    });

    describe("GIVEN member can not be found", () => {
        const interaction = {
            guild: {
                members: {
                    cache: {
                        find: jest.fn().mockReturnValue(undefined),
                    },
                    fetch: jest.fn().mockResolvedValue(undefined),
                },
            },
            reply: jest.fn(),
            update: jest.fn(),
            customId: "moons list userId 0",
        } as unknown as ButtonInteraction;

        beforeAll(async () => {
            UserSetting.FetchOneByKey = jest.fn();
            Moon.FetchPaginatedMoonsByUserId = jest.fn().mockResolvedValue(undefined)

            await List(interaction);
        });

        test("EXPECT API to be called", () => {
            expect(interaction.guild?.members.fetch).toHaveBeenCalledTimes(1);
            expect(interaction.guild?.members.fetch).toHaveBeenCalledWith("userId");
        });

        test("EXPECT error replied with username", () => {
            expect(interaction.reply).toHaveBeenCalledTimes(1);
            expect(interaction.reply).toHaveBeenCalledWith("This user does not have any moons or page is invalid.");
        });
    });
});

describe("GIVEN no moons on current page", () => {
    let updatedWith: EmbedBuilder[] | undefined;

    const interaction = {
        guild: {
            members: {
                cache: {
                    find: jest.fn().mockReturnValue({
                        user: {
                            username: "username",
                        },
                    }),
                },
            },
        },
        reply: jest.fn(),
        update: jest.fn((options: any) => {
            updatedWith = options.embeds;
        }),
        customId: "moons list userId 0",
    } as unknown as ButtonInteraction;

    beforeAll(async () => {
        UserSetting.FetchOneByKey = jest.fn();
        Moon.FetchPaginatedMoonsByUserId = jest.fn().mockResolvedValue([
            [],
            0,
        ]);

        await List(interaction);
    });

    test("EXPECT description to say so", () => {
        expect(updatedWith).toBeDefined();
        expect(updatedWith?.length).toBe(1);

        expect(updatedWith![0].data.description).toBe("*none*");
    });
});