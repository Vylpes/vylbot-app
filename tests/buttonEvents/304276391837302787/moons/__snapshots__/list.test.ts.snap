// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GIVEN happy flow EXPECT embed to be updated 1`] = `
[
  {
    "color": 3166394,
    "description": "**1 -** Test Descriptio",
    "footer": {
      "icon_url": undefined,
      "text": "Page 1 of 1 Â· 0 moons",
    },
    "title": "username's Moons",
  },
]
`;

exports[`GIVEN happy flow EXPECT row to be updated 1`] = `
[
  {
    "components": [
      {
        "custom_id": "moons list userId -1",
        "disabled": true,
        "emoji": undefined,
        "label": "Previous",
        "style": 1,
        "type": 2,
      },
      {
        "custom_id": "moons list userId 1",
        "disabled": true,
        "emoji": undefined,
        "label": "Next",
        "style": 1,
        "type": 2,
      },
    ],
    "type": 1,
  },
]
`;
